File: test/aidbox_sdk/snapshots/typescripthl7-fhir-r4-core/ExplanationOfBenefit.ts
import { Address } from "./Address";
import { Attachment } from "./Attachment";
import { Period } from "./Period";
import { CodeableConcept } from "./CodeableConcept";
import { Coding } from "./Coding";
import { Quantity } from "./Quantity";
import { DomainResource } from "./DomainResource";
import { Money } from "./Money";
import { Reference } from "./Reference";
import { Identifier } from "./Identifier";
import { BackboneElement } from "./BackboneElement";

export type ExplanationOfBenefitInsurance = BackboneElement & {
    focal: boolean;
    coverage: Reference;
    preAuthRef?: string[];
};

export type ExplanationOfBenefitBenefitBalanceFinancial = BackboneElement & {
    type: CodeableConcept;
    used?: Money | number;
    allowed?: Money | string | number;
};

export type ExplanationOfBenefitBenefitBalance = BackboneElement & {
    name?: string;
    term?: CodeableConcept;
    unit?: CodeableConcept;
    network?: CodeableConcept;
    category: CodeableConcept;
    excluded?: boolean;
    financial?: ExplanationOfBenefitBenefitBalanceFinancial[];
    description?: string;
};

export type ExplanationOfBenefitProcessNote = BackboneElement & {
    text?: string;
    type?: string;
    number?: number;
    language?: CodeableConcept;
};

export type ExplanationOfBenefitDiagnosis = BackboneElement & {
    type?: CodeableConcept[];
    sequence: number;
    diagnosis?: Reference | CodeableConcept;
    onAdmission?: CodeableConcept;
    packageCode?: CodeableConcept;
};

export type ExplanationOfBenefitSupportingInfo = BackboneElement & {
    category: CodeableConcept;
    value?: Reference | Quantity | string | boolean | Attachment;
    reason?: Coding;
    timing?: Period | string;
    sequence: number;
    code?: CodeableConcept;
};

export type ExplanationOfBenefitPayment = BackboneElement & {
    date?: string;
    type?: CodeableConcept;
    amount?: Money;
    adjustment?: Money;
    identifier?: Identifier;
    adjustmentReason?: CodeableConcept;
};

export type ExplanationOfBenefitItemAdjudication = BackboneElement & {
    value?: number;
    amount?: Money;
    reason?: CodeableConcept;
    category: CodeableConcept;
};

export type ExplanationOfBenefitItemDetailSubDetail = BackboneElement & {
    category?: CodeableConcept;
    modifier?: CodeableConcept[];
    revenue?: CodeableConcept;
    adjudication?: Reference[];
    net?: Money;
    productOrService: CodeableConcept;
    udi?: Reference[];
    programCode?: CodeableConcept[];
    factor?: number;
    sequence: number;
    quantity?: Quantity;
    noteNumber?: number[];
    unitPrice?: Money;
};

export type ExplanationOfBenefitItemDetail = BackboneElement & {
    category?: CodeableConcept;
    modifier?: CodeableConcept[];
    revenue?: CodeableConcept;
    adjudication?: Reference[];
    net?: Money;
    productOrService: CodeableConcept;
    udi?: Reference[];
    programCode?: CodeableConcept[];
    factor?: number;
    sequence: number;
    subDetail?: ExplanationOfBenefitItemDetailSubDetail[];
    quantity?: Quantity;
    noteNumber?: number[];
    unitPrice?: Money;
};

export type ExplanationOfBenefitItem = BackboneElement & {
    category?: CodeableConcept;
    diagnosisSequence?: number[];
    procedureSequence?: number[];
    modifier?: CodeableConcept[];
    revenue?: CodeableConcept;
    adjudication?: ExplanationOfBenefitItemAdjudication[];
    encounter?: Reference[];
    net?: Money;
    serviced?: string | Period;
    subSite?: CodeableConcept[];
    careTeamSequence?: number[];
    productOrService: CodeableConcept;
    udi?: Reference[];
    informationSequence?: number[];
    programCode?: CodeableConcept[];
    factor?: number;
    sequence: number;
    bodySite?: CodeableConcept;
    quantity?: Quantity;
    location?: Address | CodeableConcept | Reference;
    noteNumber?: number[];
    unitPrice?: Money;
    detail?: ExplanationOfBenefitItemDetail[];
};

export type ExplanationOfBenefitProcedure = BackboneElement & {
    udi?: Reference[];
    date?: string;
    type?: CodeableConcept[];
    sequence: number;
    procedure?: Reference | CodeableConcept;
};

export type ExplanationOfBenefitRelated = BackboneElement & {
    claim?: Reference;
    reference?: Identifier;
    relationship?: CodeableConcept;
};

export type ExplanationOfBenefitTotal = BackboneElement & {
    amount: Money;
    category: CodeableConcept;
};

export type ExplanationOfBenefitAccident = BackboneElement & {
    date?: string;
    type?: CodeableConcept;
    location?: Address | Reference;
};

export type ExplanationOfBenefitPayee = BackboneElement & {
    type?: CodeableConcept;
    party?: Reference;
};

export type ExplanationOfBenefitAddItemDetailSubDetail = BackboneElement & {
    net?: Money;
    factor?: number;
    modifier?: CodeableConcept[];
    quantity?: Quantity;
    unitPrice?: Money;
    noteNumber?: number[];
    adjudication?: Reference[];
    productOrService: CodeableConcept;
};

export type ExplanationOfBenefitAddItemDetail = BackboneElement & {
    modifier?: CodeableConcept[];
    adjudication?: Reference[];
    net?: Money;
    productOrService: CodeableConcept;
    factor?: number;
    subDetail?: ExplanationOfBenefitAddItemDetailSubDetail[];
    quantity?: Quantity;
    noteNumber?: number[];
    unitPrice?: Money;
};

export type ExplanationOfBenefitAddItem = BackboneElement & {
    modifier?: CodeableConcept[];
    adjudication?: Reference[];
    itemSequence?: number[];
    net?: Money;
    serviced?: string | Period;
    detailSequence?: number[];
    subSite?: CodeableConcept[];
    productOrService: CodeableConcept;
    programCode?: CodeableConcept[];
    factor?: number;
    subDetailSequence?: number[];
    bodySite?: CodeableConcept;
    quantity?: Quantity;
    location?: Address | CodeableConcept | Reference;
    provider?: Reference[];
    noteNumber?: number[];
    unitPrice?: Money;
    detail?: ExplanationOfBenefitAddItemDetail[];
};

export type ExplanationOfBenefitCareTeam = BackboneElement & {
    role?: CodeableConcept;
    provider: Reference;
    sequence: number;
    responsible?: boolean;
    qualification?: CodeableConcept;
};

export type ExplanationOfBenefit = DomainResource & {
    patient: Reference;
    claimResponse?: Reference;
    insurance: ExplanationOfBenefitInsurance[];
    benefitBalance?: ExplanationOfBenefitBenefitBalance[];
    facility?: Reference;
    processNote?: ExplanationOfBenefitProcessNote[];
    diagnosis?: ExplanationOfBenefitDiagnosis[];
    preAuthRef?: string[];
    adjudication?: Reference[];
    enterer?: Reference;
    supportingInfo?: ExplanationOfBenefitSupportingInfo[];
    use: string;
    payment?: ExplanationOfBenefitPayment;
    item?: ExplanationOfBenefitItem[];
    type: CodeableConcept;
    created: string;
    procedure?: ExplanationOfBenefitProcedure[];
    outcome: string;
    related?: ExplanationOfBenefitRelated[];
    disposition?: string;
    referral?: Reference;
    preAuthRefPeriod?: Period[];
    total?: ExplanationOfBenefitTotal[];
    insurer: Reference;
    fundsReserve?: CodeableConcept;
    priority?: CodeableConcept;
    accident?: ExplanationOfBenefitAccident;
    status: string;
    payee?: ExplanationOfBenefitPayee;
    prescription?: Reference;
    billablePeriod?: Period;
    identifier?: Identifier[];
    form?: Attachment;
    subType?: CodeableConcept;
    fundsReserveRequested?: CodeableConcept;
    benefitPeriod?: Period;
    precedence?: number;
    formCode?: CodeableConcept;
    provider: Reference;
    addItem?: ExplanationOfBenefitAddItem[];
    originalPrescription?: Reference;
    careTeam?: ExplanationOfBenefitCareTeam[];
    claim?: Reference;
};

File: test/aidbox_sdk/snapshots/python/hl7_fhir_r4_core/ExplanationOfBenefit.py
from typing import Optional, List
from dataclasses import dataclass, field
from base import Address
from base import Attachment
from base import Period
from base import CodeableConcept
from base import Coding
from base import Quantity
from base import DomainResource
from base import Money
from base import Reference
from base import Identifier
from base import BackboneElement

@dataclass(kw_only=True)
class ExplanationOfBenefitInsurance(BackboneElement):
    focal: bool
    coverage: Reference
    pre_auth_ref: Optional[List[str]] = field(default_factory=list)

@dataclass(kw_only=True)
class ExplanationOfBenefitBenefitBalanceFinancial(BackboneElement):
    type: CodeableConcept
    used_money: Optional[Money] = None
    allowed_money: Optional[Money] = None
    allowed_string: Optional[str] = None
    used_unsigned_int: Optional[int] = None
    allowed_unsigned_int: Optional[int] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitBenefitBalance(BackboneElement):
    name: Optional[str] = None
    term: Optional[CodeableConcept] = None
    unit: Optional[CodeableConcept] = None
    network: Optional[CodeableConcept] = None
    category: CodeableConcept
    excluded: Optional[bool] = None
    financial: Optional[List[ExplanationOfBenefitBenefitBalanceFinancial]] = field(default_factory=list)
    description: Optional[str] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitProcessNote(BackboneElement):
    text: Optional[str] = None
    type: Optional[str] = None
    number: Optional[int] = None
    language: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitDiagnosis(BackboneElement):
    type: Optional[List[CodeableConcept]] = field(default_factory=list)
    sequence: int
    on_admission: Optional[CodeableConcept] = None
    package_code: Optional[CodeableConcept] = None
    diagnosis_reference: Optional[Reference] = None
    diagnosis_codeable_concept: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitSupportingInfo(BackboneElement):
    category: CodeableConcept
    value_reference: Optional[Reference] = None
    value_quantity: Optional[Quantity] = None
    timing_period: Optional[Period] = None
    value_string: Optional[str] = None
    value_boolean: Optional[bool] = None
    reason: Optional[Coding] = None
    sequence: int
    code: Optional[CodeableConcept] = None
    timing_date: Optional[str] = None
    value_attachment: Optional[Attachment] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitPayment(BackboneElement):
    date: Optional[str] = None
    type: Optional[CodeableConcept] = None
    amount: Optional[Money] = None
    adjustment: Optional[Money] = None
    identifier: Optional[Identifier] = None
    adjustment_reason: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitItemAdjudication(BackboneElement):
    value: Optional[float] = None
    amount: Optional[Money] = None
    reason: Optional[CodeableConcept] = None
    category: CodeableConcept

@dataclass(kw_only=True)
class ExplanationOfBenefitItemDetailSubDetail(BackboneElement):
    category: Optional[CodeableConcept] = None
    modifier: Optional[List[CodeableConcept]] = field(default_factory=list)
    revenue: Optional[CodeableConcept] = None
    adjudication: Optional[List[Reference]] = field(default_factory=list)
    net: Optional[Money] = None
    product_or_service: CodeableConcept
    udi: Optional[List[Reference]] = field(default_factory=list)
    program_code: Optional[List[CodeableConcept]] = field(default_factory=list)
    factor: Optional[float] = None
    sequence: int
    quantity: Optional[Quantity] = None
    note_number: Optional[List[int]] = field(default_factory=list)
    unit_price: Optional[Money] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitItemDetail(BackboneElement):
    category: Optional[CodeableConcept] = None
    modifier: Optional[List[CodeableConcept]] = field(default_factory=list)
    revenue: Optional[CodeableConcept] = None
    adjudication: Optional[List[Reference]] = field(default_factory=list)
    net: Optional[Money] = None
    product_or_service: CodeableConcept
    udi: Optional[List[Reference]] = field(default_factory=list)
    program_code: Optional[List[CodeableConcept]] = field(default_factory=list)
    factor: Optional[float] = None
    sequence: int
    sub_detail: Optional[List[ExplanationOfBenefitItemDetailSubDetail]] = field(default_factory=list)
    quantity: Optional[Quantity] = None
    note_number: Optional[List[int]] = field(default_factory=list)
    unit_price: Optional[Money] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitItem(BackboneElement):
    category: Optional[CodeableConcept] = None
    diagnosis_sequence: Optional[List[int]] = field(default_factory=list)
    procedure_sequence: Optional[List[int]] = field(default_factory=list)
    location_address: Optional[Address] = None
    modifier: Optional[List[CodeableConcept]] = field(default_factory=list)
    revenue: Optional[CodeableConcept] = None
    adjudication: Optional[List[ExplanationOfBenefitItemAdjudication]] = field(default_factory=list)
    encounter: Optional[List[Reference]] = field(default_factory=list)
    location_codeable_concept: Optional[CodeableConcept] = None
    net: Optional[Money] = None
    sub_site: Optional[List[CodeableConcept]] = field(default_factory=list)
    care_team_sequence: Optional[List[int]] = field(default_factory=list)
    product_or_service: CodeableConcept
    location_reference: Optional[Reference] = None
    udi: Optional[List[Reference]] = field(default_factory=list)
    information_sequence: Optional[List[int]] = field(default_factory=list)
    program_code: Optional[List[CodeableConcept]] = field(default_factory=list)
    factor: Optional[float] = None
    serviced_date: Optional[str] = None
    sequence: int
    body_site: Optional[CodeableConcept] = None
    quantity: Optional[Quantity] = None
    note_number: Optional[List[int]] = field(default_factory=list)
    unit_price: Optional[Money] = None
    serviced_period: Optional[Period] = None
    detail: Optional[List[ExplanationOfBenefitItemDetail]] = field(default_factory=list)

@dataclass(kw_only=True)
class ExplanationOfBenefitProcedure(BackboneElement):
    udi: Optional[List[Reference]] = field(default_factory=list)
    date: Optional[str] = None
    type: Optional[List[CodeableConcept]] = field(default_factory=list)
    sequence: int
    procedure_reference: Optional[Reference] = None
    procedure_codeable_concept: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitRelated(BackboneElement):
    claim: Optional[Reference] = None
    reference: Optional[Identifier] = None
    relationship: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitTotal(BackboneElement):
    amount: Money
    category: CodeableConcept

@dataclass(kw_only=True)
class ExplanationOfBenefitAccident(BackboneElement):
    date: Optional[str] = None
    type: Optional[CodeableConcept] = None
    location_address: Optional[Address] = None
    location_reference: Optional[Reference] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitPayee(BackboneElement):
    type: Optional[CodeableConcept] = None
    party: Optional[Reference] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitAddItemDetailSubDetail(BackboneElement):
    net: Optional[Money] = None
    factor: Optional[float] = None
    modifier: Optional[List[CodeableConcept]] = field(default_factory=list)
    quantity: Optional[Quantity] = None
    unit_price: Optional[Money] = None
    note_number: Optional[List[int]] = field(default_factory=list)
    adjudication: Optional[List[Reference]] = field(default_factory=list)
    product_or_service: CodeableConcept

@dataclass(kw_only=True)
class ExplanationOfBenefitAddItemDetail(BackboneElement):
    modifier: Optional[List[CodeableConcept]] = field(default_factory=list)
    adjudication: Optional[List[Reference]] = field(default_factory=list)
    net: Optional[Money] = None
    product_or_service: CodeableConcept
    factor: Optional[float] = None
    sub_detail: Optional[List[ExplanationOfBenefitAddItemDetailSubDetail]] = field(default_factory=list)
    quantity: Optional[Quantity] = None
    note_number: Optional[List[int]] = field(default_factory=list)
    unit_price: Optional[Money] = None

@dataclass(kw_only=True)
class ExplanationOfBenefitAddItem(BackboneElement):
    location_address: Optional[Address] = None
    modifier: Optional[List[CodeableConcept]] = field(default_factory=list)
    adjudication: Optional[List[Reference]] = field(default_factory=list)
    location_codeable_concept: Optional[CodeableConcept] = None
    item_sequence: Optional[List[int]] = field(default_factory=list)
    net: Optional[Money] = None
    detail_sequence: Optional[List[int]] = field(default_factory=list)
    sub_site: Optional[List[CodeableConcept]] = field(default_factory=list)
    product_or_service: CodeableConcept
    location_reference: Optional[Reference] = None
    program_code: Optional[List[CodeableConcept]] = field(default_factory=list)
    factor: Optional[float] = None
    serviced_date: Optional[str] = None
    sub_detail_sequence: Optional[List[int]] = field(default_factory=list)
    body_site: Optional[CodeableConcept] = None
    quantity: Optional[Quantity] = None
    provider: Optional[List[Reference]] = field(default_factory=list)
    note_number: Optional[List[int]] = field(default_factory=list)
    unit_price: Optional[Money] = None
    serviced_period: Optional[Period] = None
    detail: Optional[List[ExplanationOfBenefitAddItemDetail]] = field(default_factory=list)

@dataclass(kw_only=True)
class ExplanationOfBenefitCareTeam(BackboneElement):
    role: Optional[CodeableConcept] = None
    provider: Reference
    sequence: int
    responsible: Optional[bool] = None
    qualification: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class ExplanationOfBenefit(DomainResource):
    patient: Reference
    claim_response: Optional[Reference] = None
    insurance: list[ExplanationOfBenefitInsurance] = field(default_factory=list)
    benefit_balance: Optional[List[ExplanationOfBenefitBenefitBalance]] = field(default_factory=list)
    facility: Optional[Reference] = None
    process_note: Optional[List[ExplanationOfBenefitProcessNote]] = field(default_factory=list)
    diagnosis: Optional[List[ExplanationOfBenefitDiagnosis]] = field(default_factory=list)
    pre_auth_ref: Optional[List[str]] = field(default_factory=list)
    adjudication: Optional[List[Reference]] = field(default_factory=list)
    enterer: Optional[Reference] = None
    supporting_info: Optional[List[ExplanationOfBenefitSupportingInfo]] = field(default_factory=list)
    use: str
    payment: Optional[ExplanationOfBenefitPayment] = None
    item: Optional[List[ExplanationOfBenefitItem]] = field(default_factory=list)
    type: CodeableConcept
    created: str
    procedure: Optional[List[ExplanationOfBenefitProcedure]] = field(default_factory=list)
    outcome: str
    related: Optional[List[ExplanationOfBenefitRelated]] = field(default_factory=list)
    disposition: Optional[str] = None
    referral: Optional[Reference] = None
    pre_auth_ref_period: Optional[List[Period]] = field(default_factory=list)
    total: Optional[List[ExplanationOfBenefitTotal]] = field(default_factory=list)
    insurer: Reference
    funds_reserve: Optional[CodeableConcept] = None
    priority: Optional[CodeableConcept] = None
    accident: Optional[ExplanationOfBenefitAccident] = None
    status: str
    payee: Optional[ExplanationOfBenefitPayee] = None
    prescription: Optional[Reference] = None
    billable_period: Optional[Period] = None
    identifier: Optional[List[Identifier]] = field(default_factory=list)
    form: Optional[Attachment] = None
    sub_type: Optional[CodeableConcept] = None
    funds_reserve_requested: Optional[CodeableConcept] = None
    benefit_period: Optional[Period] = None
    precedence: Optional[int] = None
    form_code: Optional[CodeableConcept] = None
    provider: Reference
    add_item: Optional[List[ExplanationOfBenefitAddItem]] = field(default_factory=list)
    original_prescription: Optional[Reference] = None
    care_team: Optional[List[ExplanationOfBenefitCareTeam]] = field(default_factory=list)
    claim: Optional[Reference] = None